Make sure Functional algorithms check types during conversion

tmadlener opened the pull request:
BEGINRELEASENOTES
- Make sure that the internal casts that happen for functional algorithms do not silently cast to wrong types.

ENDRELEASENOTES

Fixes #274 

- [x] which other `static_cast`s need these checks?
- [x] More tests, e.g. at least one algorithm where the branch with the vector of inputs and multiple inputs is triggered.

tmadlener commented:
From my point of view this is ready for review. I have put two comments / questions for clarifications inline.

m-fila commented on k4FWCore/include/k4FWCore/FunctionalUtils.h:
```diff
@@ -136,15 +137,35 @@ namespace k4FWCore {
+                    thisClass->name(),
+                    fmt::format(
+                        "Failed to cast collection {} to the required type {}, the type of the collection is {}",
+                        handle.objKey(), typeid(EDM4hepType).name(), in ? in->getTypeName() : "[undetermined]"),
```
> the `EDM4hepType::typeName` fails here because `EDM4hepType` can still be a `podio::CollectionBase`

Is that right? The exception is thrown in the `else` branch of `if constexpr (std::is_same_v<EDM4hepType, const podio::CollectionBase*>)`, no?

tmadlener replied:
Looking at this again, it should in principle work, but I ran into a compiler error here. The compiler error comes from (as far as I can tell) the `CollectionMerger` which operates on a vector of `CollectionBase`:

https://github.com/key4hep/k4FWCore/blob/ce56844f12a0bab64954604466e780d63e021ed1/k4FWCore/components/CollectionMerger.cpp#L45

the issue is that the `if constexpr` checks for a `podio::CollectionBase*`, but in this specific instantiation, `EDM4hepType` is a `podio::CollectionBase` (**without pointer**). I haven't fully understood yet, whey the `remove_pointer_t` even makes it possible to still have a `podio::CollectionBase*` at that point, though.

tmadlener replied:
OK. That `if constexpr` branch is never taken. I have no idea, how the collection merger even works at this point.

jmcarcell replied:
See https://github.com/key4hep/k4FWCore/pull/275/files#r1936892322

tmadlener commented on k4FWCore/include/k4FWCore/FunctionalUtils.h:
```diff
@@ -136,15 +139,35 @@ namespace k4FWCore {
+              podio::CollectionBase* in      = handle.get()->get();
+              auto*                  typedIn = dynamic_cast<EDM4hepType*>(in);
+              if (typedIn) {
+                inputMap.push_back(static_cast<EDM4hepType*>(typedIn));
```
```suggestion
                inputMap.push_back(typedIn);
```
No need to cast again, this already has the correct type.

tmadlener commented:
I have reworked the logic in `readVectorInputs` such that the `if constexpr` branch is actually taken by the `CollectionMerger` (that operates on `podio::CollectionBase*`). That makes it possible to use the `typeName` static member to improve the output here.

I have also introduced a `TupleType` typedef to make the implementation a bit more readable, added a `remove_cvref` to also remove spurious `const`s and explicitly add them back where necessary.

The tests are still passing, so I suppose this is should not break other things downstream, but functional algorithms that work on `podio::CollectionBase*` will take a different branch now than they did before.

jmcarcell commented:
I've split the changes not related to the cast in https://github.com/key4hep/k4FWCore/pull/276. I've realised I also need changes in that function for mixing functional algorithms and Marlin processors so it's going to be very messy if not split.

jmcarcell commented on k4FWCore/include/k4FWCore/FunctionalUtils.h:
```diff
@@ -133,16 +136,34 @@ namespace k4FWCore {
+              throw GaudiException(
+                  thisClass->name(),
+                  fmt::format("Failed to cast collection {} to the required type {}, the type of the collection is {}",
+                              handle.objKey(), typeid(EDM4hepType).name(), in ? in->getTypeName() : "[undetermined]"),
```
Because of the if - else that was removed in #276 `typeid` has to be used here... I think it's acceptable considering this will only happen for when reading using an `std::vector` unlike the usual normal input below. It looks much cleaner without and if - else to check if it's a `podio::CollectionBase`.

tmadlener replied:
Yeah, I think it's OK. If people really need more we can always make it better later.

jmcarcell commented on k4FWCore/include/k4FWCore/FunctionalUtils.h:
```diff
@@ -136,15 +137,35 @@ namespace k4FWCore {
             } else {
-              podio::CollectionBase* in = handle.get()->get();
-              inputMap.push_back(static_cast<EDM4hepType*>(in));
+              podio::CollectionBase* in      = handle.get()->get();
```
I think this is outdated now since both ways are handled the same way

tmadlener replied:
Yes, I think you fixed that when reworking the special handling of the `podio::CollectionBase` (resp. removing that).

jmcarcell commented:
Waiting for @m-fila to say something or :+1: to merge.

m-fila commented:
> Waiting for @m-fila to say something or 👍 to merge.

Sorry, please give me some a little more time to check again

m-fila commented on k4FWCore/include/k4FWCore/FunctionalUtils.h:
```diff
@@ -133,16 +136,34 @@ namespace k4FWCore {
+              throw GaudiException(
+                  thisClass->name(),
+                  fmt::format("Failed to cast collection {} to the required type {}, the type of the collection is {}",
+                              std::get<Index>(handles)[0].objKey(), EDM4hepType::typeName, in->getTypeName()),
```
Since `in` is a pointer that was just obtained shouldn't there be also a check for it's validity - just like in the other branch?

tmadlener replied:
Good catch. Done

jmcarcell replied:
I believe it's almost impossible for `in` to be a null pointer since with the current interface algorithms only give valid collections as output. For it to happen it should be in an algorithm like `CollectionMerger.cpp` where pointers are given directly in the output.

m-fila reviewed (approved):
I have a single question, otherwise looks good
